Issue #24534: Msg for wrong package Resource annotation

tkburroughs opened the pull request:
Log an audit message when a Jakarta EE 9+ application includes use of javax.annotation.Resource, javax.annotation.PostConstruct, or javax.annotation.PreDestroy

for #24534

tkburroughs committed:
Issue #24535: Msg for wrong package Resource annotation
Log an audit message when a Jakarta EE 9+ application includes use of
javax.annotation.Resource, javax.annotation.PostConstruct, or javax.annotation.PreDestroy

tkburroughs commented:
#build

posmith commented:
L2 message review completed

helyarp commented on dev/com.ibm.ws.container.service/resources/com/ibm/ws/container/service/resources/Messages.nlsprops:
```diff
@@ -166,4 +166,6 @@ ordered.list.already.contains.CWWKM0482E=CWWKM0482E: WebSphere Application Serve
 ordered.list.already.contains.CWWKM0482E.useraction=See the problem determination information on the WebSphere Application Server Support web page: http://www.ibm.com/software/webservers/appserv/was/support.
 
-
+wrong.annotation.package.CWWKM0483I=CWWKM0483I: The {0} annotation will be ignored on classes in the {1} module. Jakarta EE applications must use the {2} annotation. The affected classes are: {3}
```
We are not supposed to use future tense in our messages. Please change "will be" to "is" unless this message requires "will be" to be technically correct.

tkburroughs replied:
done

helyarp commented on dev/com.ibm.ws.container.service/resources/com/ibm/ws/container/service/resources/Messages.nlsprops:
```diff
@@ -166,4 +166,6 @@ ordered.list.already.contains.CWWKM0482E=CWWKM0482E: WebSphere Application Serve
 
-
+wrong.annotation.package.CWWKM0483I=CWWKM0483I: The {0} annotation will be ignored on classes in the {1} module. Jakarta EE applications must use the {2} annotation. The affected classes are: {3}
+wrong.annotation.package.CWWKM0483I.explanation=The application contains a Java EE annotation with the 'javax' package name, but the application server has Jakarta EE features enabled. The application server will not process the 'javax' package name annotation.
```
Please change "will" to "did" unless "will" is required to make the message technically correct.

tkburroughs replied:
done

helyarp reviewed (commented):
Suggested changes

tkburroughs commented:
@helyarp changes made as suggested, thanks.

tkburroughs commented:
#libby
#build

tkburroughs commented:
#build

tkburroughs commented:
Latest build with message improvements was good; ready to merge.

helyarp reviewed (commented):
Looks good!

olendvcook commented on dev/com.ibm.ws.container.service/src/com/ibm/ws/container/service/annocache/internal/AnnotationsImpl.java:
```diff
@@ -735,7 +744,76 @@ protected void addExternalToClassSource() {
+            return false;
+        }
+        try {
+            classLoader.loadClass("jakarta.annotation.Resource");
```
Could be replaced with JAKARTA_RESOURCE variable

tkburroughs replied:
changed

olendvcook commented on dev/com.ibm.ws.container.service/src/com/ibm/ws/container/service/annocache/internal/AnnotationsImpl.java:
```diff
@@ -735,7 +744,76 @@ protected void addExternalToClassSource() {
+
+        Set<String> foundClasses = new TreeSet<String>(); // provides natural ordering
+
+        // Look for @Resoruce at class, field, and method level and nested in @Resoruces at class level
```
"@Resoruce" - spelled wrong in the comment

tkburroughs replied:
fixed

olendvcook commented on dev/io.openliberty.injection.jakarta_fat/fat/src/io/openliberty/injection/jakarta/fat/tests/WrongPackageTest.java:
```diff
@@ -0,0 +1,128 @@
+                         ee9server.findStringsInLogsUsingMark(CWWKM0483I_EJB_MOD_RESOURCE, ee9server.getDefaultLogFile()).size());
+            assertEquals("Expected CWWKM0483I message not found for @PreDestroy in EJB module : " + CWWKM0483I_EJB_MOD_PREDESTROY, 1,
+                         ee9server.findStringsInLogsUsingMark(CWWKM0483I_EJB_MOD_PREDESTROY, ee9server.getDefaultLogFile()).size());
+            assertEquals("Expected CWWKM0483I message not found for @Resoruce in WEB module : " + CWWKM0483I_WEB_MOD_RESOURCE, 1,
```
@Resoruce again

tkburroughs replied:
fixed

olendvcook commented on dev/io.openliberty.injection.jakarta_fat/fat/src/io/openliberty/injection/jakarta/fat/tests/WrongPackageTest.java:
```diff
@@ -0,0 +1,128 @@
+                         ee10server.findStringsInLogsUsingMark(CWWKM0483I_EJB_MOD_RESOURCE, ee10server.getDefaultLogFile()).size());
+            assertEquals("Expected CWWKM0483I message not found for @PreDestroy in EJB module : " + CWWKM0483I_EJB_MOD_PREDESTROY, 1,
+                         ee10server.findStringsInLogsUsingMark(CWWKM0483I_EJB_MOD_PREDESTROY, ee10server.getDefaultLogFile()).size());
+            assertEquals("Expected CWWKM0483I message not found for @Resoruce in WEB module : " + CWWKM0483I_WEB_MOD_RESOURCE, 1,
```
@Resoruce again

tkburroughs replied:
fixed

olendvcook commented on dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageWeb.war/src/io/openliberty/injection/jakarta/web/WrongPackageServlet.java:
```diff
@@ -0,0 +1,83 @@
+     * warning per module. The jakarta.annotation.Resource annotation works as expected.
+     */
+    public void testWrongPackageCommonAnnotations() throws Exception {
+        assertNotNull("jakarta Resrouce UserTransaction is null", userTx);
```
Resrouce

tkburroughs replied:
fixed

olendvcook commented:
~~I'm finding it hard to google if Jakarta @Resources is a valid annotation, or if they scrapped it, but I guess I'm assuming based on these changes that it doesn't exist, since this seems to only mention javax @Resources.~~

Nevermind, Jakarta @Resources does exist but the code here is grouping it up with the jakarta Resource message.

tkburroughs commented:
#libby
#build

tkburroughs committed:
Issue #24534: Msg for wrong package resource annotation
- add app name to message and update test to include web app

tkburroughs commented:
@helyarp At design review it was requested I update the message to include the application name, so I've updated to add that.  Please review message gain... Thanks.

tkburroughs commented:
#libby
#build

tkburroughs commented:
@olendvcook per design review, I've
- updated message to include the application name (and updated code)
- updated the test to check for correct application name in msg
- updated the test to include standalone web application (war) to ensure proper application name

Please review again... thanks.

helyarp reviewed (commented):
Message changes look good!

olendvcook commented on dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageEJB.jar/src/io/openliberty/injection/jakarta/ejb/JakartaSingletonResourcesBean.java:
```diff
@@ -0,0 +1,45 @@
+        try {
+            Context namingCtx = new InitialContext();
+            SessionContext sctx = (SessionContext) namingCtx.lookup("java:comp/env/SessionContext");
+            fail("lookup of javax Resrouce was successful : " + sctx);
```
Resrouce

tkburroughs replied:
fixed

olendvcook commented on dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageEJB.jar/src/io/openliberty/injection/jakarta/ejb/JakartaStatelessBean.java:
```diff
@@ -0,0 +1,41 @@
+    JakartaSingletonBean bean;
+
+    public void verifyInjection() {
+        assertNotNull("jakarta Resrouce EJBContext is null", ctx);
```
Resrouce

tkburroughs replied:
fixed

olendvcook commented on dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageWeb.war/src/io/openliberty/injection/jakarta/web/JakartaSingletonWarBean.java:
```diff
@@ -0,0 +1,41 @@
+    JakartaStatelessWarBean bean;
+
+    public void verifyInjection() {
+        assertNotNull("jakarta Resrouce EJBContext is null", ctx);
```
Resrouce

tkburroughs replied:
fixed

olendvcook commented on dev/io.openliberty.injection.jakarta_fat/test-applications/WrongPackageWeb.war/src/io/openliberty/injection/jakarta/web/JakartaStatelessWarBean.java:
```diff
@@ -0,0 +1,41 @@
+    JakartaSingletonWarBean bean;
+
+    public void verifyInjection() {
+        assertNotNull("Jakarta Resrouce EJBContext is null", ctx);
```
Resrouce

tkburroughs replied:
fixed

olendvcook commented:
I found some more misspellings that I missed the first review.

tkburroughs commented:
@olendvcook Fixed.

tkburroughs commented:
#libby
#build

tkburroughs commented:
All test failures in build linked to existing know defects; ready to merge.
